### PR TITLE
content/assignments: Add "Lambda Function Loader" checker

### PR DESCRIPTION
**Pull request by fabianpatras**
Add `tests/` directory.
In oreder to run the checker, `cd` into `tests` and run `make check`.

# Prerequisite Checklist

- [x] Read the [contribution guidelines](https://github.com/cs-pub-ro/operating-systems/blob/main/CONTRIBUTING.md#pull-requests) regarding submitting new changes to the project;
- [x] Tested your changes against relevant architectures and platforms;
- [x] Updated relevant documentation (if needed).

## Description of changes

This PR adds `tests` to the Lambda Loader assignment.

TODO:
- [DONE] update assignment's `README.md` file to reflect the new rules on how to run the checker.

### CODE REVIEW

File: content/assignments/lambda-function-loader/tests/Makefile [hunk 1]
```diff
@@ -0,0 +1,31 @@
+SRCDIR?=$(abspath ../src)
+CC=gcc
+CFLAGS=-fPIC -Wall -Werror
+LDLIBS=$(SRCDIR)/libipc.so
+CLIENT_LIBS=libbasic.so libadvanced.so libspecial.so
+
+.PHONY: build
```
Comment (teodutu) [inline content/assignments/lambda-function-loader/tests/Makefile:7]:
```suggestion
.PHONY: build check clean
```

File: content/assignments/lambda-function-loader/tests/Makefile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+build: client $(CLIENT_LIBS)
+
+.PHONY: check
```
Comment (teodutu) [inline content/assignments/lambda-function-loader/tests/Makefile:10]:
```suggestion
```

File: content/assignments/lambda-function-loader/tests/Makefile [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+$(LDLIBS):
+	@$(MAKE) -C $(SRCDIR) -f Makefile $(foreach LIB,$(LDLIBS),$(notdir $(LIB)))
+
+lib%.so: %.o
+	$(CC) $(LDFLAGS) -shared -o $@ $^
+
+%.o: %.c
+	$(CC) $(CFLAGS) -o $@ -c $<
+
+.PHONY: clean
```
Comment (teodutu) [inline content/assignments/lambda-function-loader/tests/Makefile:28]:
```suggestion
```

**Review by teodutu — CHANGES_REQUESTED** [review]
I just made a brief suggestion without having tested the checker yet

File: content/assignments/lambda-function-loader/README.md [hunk 4]
```diff
@@ -104,28 +104,36 @@ What is needed for the implementation of the proposed project:
...
-student@os:~/checker$ ./client $(realpath libbasic.so)
-Output file: /home/so/checker/output/out-cfy0fl
-
-student@os:~/checker$ cat /home/so/output/out-cfy0fl
+#
+# Open another terminal and instruct the client to send a library to the server.
+# Please remember the calling convention for client
+# ./client <libname> [<function> [<file>]]
+#
+# Note: When <function> is not provided, the server will try to execute the `run` function.
```
Comment (elena19m) [inline content/assignments/lambda-function-loader/README.md:116]:
```suggestion
# Note: When <function> is not provided, the server must execute the `run` function.
```

File: content/assignments/lambda-function-loader/README.md [hunk 5]
```diff
@@ -104,28 +104,36 @@ What is needed for the implementation of the proposed project:
...
+student@os:../lambda-function-loader/$ cd tests
+student@os:../lambda-function-loader/tests$ make
+student@os:../lambda-function-loader/tests$ ./client $(realpath libbasic.so)
+Output file: /home/student/operating-systems/content/assignments/lambda-function-loader/tests/output/out-bSJdTv
+
+student@os:../lambda-function-loader/tests$ cat /home/student/operating-systems/content/assignments/lambda-function-loader/tests/output/out-bSJdTv
 run
 
 # Execute the "function" function from libbasic.so
```
Comment (elena19m) [inline content/assignments/lambda-function-loader/README.md:125]:
```suggestion
# Execute the "function" function from libbasic.so. The function does not exist, so an error is printed instead.
```

File: content/assignments/lambda-function-loader/README.md [hunk 6]
```diff
@@ -104,28 +104,36 @@ What is needed for the implementation of the proposed project:
...
 
-student@os:~/checker$ cat /home/so/output/out-vc7s03
-function
+student@os:../lambda-function-loader/tests$ cat /home/so/output/out-qOcoAA
+Error: /home/student/operating-systems/content/assignments/lambda-function-loader/tests/libbasic.so function could not be executed.
 
-# Execute the "cat" function from libbasic.so with the argument "/home/so/checker/Makefile"
-student@os:~/checker$ ./client $(realpath libbasic.so) cat $(realpath Makefile)
-Output file: /home/so/checker/output/out-y732bN
+# Execute the "cat" function from libbasic.so with the argumentbeing the full path of file "Makefile"
```
Comment (elena19m) [inline content/assignments/lambda-function-loader/README.md:132]:
```suggestion
# Execute the "cat" function from libbasic.so with the argument being the full path of file "Makefile"
```

File: content/assignments/lambda-function-loader/README.md [hunk 7]
```diff
@@ -134,10 +142,30 @@ CC=gcc
 
 You have a checker available for partial verification of your implementation.
 
-```text
-student@os~$ cd checker
+The checker behaves as follows:
+- it automatically builds the source files then it starts the `server`
+- it runs all the tests without restarting the `server`
+- it then kills the `server`
```
Comment (elena19m) [inline content/assignments/lambda-function-loader/README.md:148]:
```suggestion
- it automatically builds the source files then it starts the `server`.
- it runs all the tests without restarting the `server`.
- it then kills the `server`.
```

**Review by elena19m — CHANGES_REQUESTED** [review]

File: content/assignments/lambda-function-loader/tests/basic.c [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+{
+	int fd;
+	char buf[1024];
+	int read_bytes;
+
+	fd = open(filename, O_RDONLY);
+
+	while ((read_bytes = read(fd, buf, sizeof(buf))) > 0)
+		write(STDOUT_FILENO, buf, read_bytes);
+
```
Comment (teodutu) [inline content/assignments/lambda-function-loader/tests/basic.c:22]:
```suggestion
```

File: content/assignments/lambda-function-loader/README.md [hunk 9]
```diff
@@ -104,28 +104,36 @@ What is needed for the implementation of the proposed project:
...
+student@os:../lambda-function-loader/tests$ ./client $(realpath libbasic.so)
+Output file: /home/student/operating-systems/content/assignments/lambda-function-loader/tests/output/out-bSJdTv
+
+student@os:../lambda-function-loader/tests$ cat /home/student/operating-systems/content/assignments/lambda-function-loader/tests/output/out-bSJdTv
 run
 
-# Execute the "function" function from libbasic.so
-student@os:~/checker$ ./client $(realpath libbasic.so) function
-Output file: /home/so/checker/output/out-vc7s03
+# Execute the "function" function from libbasic.so. The function does not exist, so an error is printed.
```
Comment (teodutu) [inline content/assignments/lambda-function-loader/README.md:125]:
```suggestion
# Execute the "function" function from libbasic.so.
# The function does not exist, so an error is printed.
```

File: content/assignments/lambda-function-loader/README.md [hunk 10]
```diff
@@ -104,28 +104,36 @@ What is needed for the implementation of the proposed project:
 
 ```bash
 ## Starting the server
-student@os:~/skel$ ./server
+student@os:../lambda-function-loader/src$ make
```
Comment (teodutu) [inline content/assignments/lambda-function-loader/README.md:107]:
```suggestion
student@os:.../lambda-function-loader/src$ make
```
`..` tends to be confused with the previous directory. `...` are generally used to symbolise a longer path. Do this for all paths.

File: content/assignments/lambda-function-loader/README.md [hunk 11]
```diff
@@ -134,10 +142,30 @@ CC=gcc
...
 You have a checker available for partial verification of your implementation.
 
-```text
-student@os~$ cd checker
+The checker behaves as follows:
+- it automatically builds the source files then it starts the `server`.
+- it runs all the tests without restarting the `server`.
+- it then kills the `server`.
+
+For debugging purposes, you can instruct the checker to restart the server before each test. Please see below.
```
Comment (teodutu) [inline content/assignments/lambda-function-loader/README.md:150]:
```suggestion
For debugging purposes, you can instruct the checker to restart the server before each test.
Please see below.
```

File: content/assignments/lambda-function-loader/README.md [hunk 12]
```diff
@@ -134,10 +142,30 @@ CC=gcc
 
 You have a checker available for partial verification of your implementation.
 
-```text
-student@os~$ cd checker
+The checker behaves as follows:
+- it automatically builds the source files then it starts the `server`.
```
Comment (teodutu) [inline content/assignments/lambda-function-loader/README.md:146]:
```suggestion

- it automatically builds the source files then it starts the `server`.
```

**Review by teodutu — CHANGES_REQUESTED** [review]
- The checkpatch errors are BS. Ignore them.
- I made inline suggestions for fixing the super-linter errors coming from `README.md`. Others such as these [1, 2, 3] and others like them can be fixed. The errors pretty much tell you how to fix them.
- Fuck the spellcheck errors. Its wordlist is way too short.

[1] https://github.com/cs-pub-ro/operating-systems/actions/runs/7226047704/job/19690782887?pr=38#step:3:701
[2] https://github.com/cs-pub-ro/operating-systems/actions/runs/7226047704/job/19690782887?pr=38#step:3:732
[3] https://github.com/cs-pub-ro/operating-systems/actions/runs/7226047704/job/19690782887?pr=38#step:3:1017

**Review by elena19m — APPROVED** [review]

**Review by teodutu — APPROVED** [review]
:+1:

**Pull request merged**